modules/kv-state-cache/ds/kv_state_cache.cc

```diff
@@ -0,0 +1,364 @@
+
+    // Transfer the data from this builder to the child builder.
+    data->offset =
+        kvStateCacheBlockBuilder->Split(childKVStateCacheBlockBuilder, index);
+  }
+  VLOG(100) << "builder:" << kvStateCacheBlockBuilder
+            << " bitmap:" << kvStateCacheBlockBuilder->GetBitmapStr();
+  VLOG(100) << "child_builder:" << childKVStateCacheBlockBuilder
+            << " bitmap:" << childKVStateCacheBlockBuilder->GetBitmapStr();
+  return childKVStateCacheBlockBuilder;
+}
+
+void KVStateCacheBuilder::Update(Client& client,
+                                 const std::vector<int>& tokenList,
+                                 int nextToken,
+                                 const KV_STATE_WITH_LAYER& kvState) {
+  std::vector<int> tokenListCopy = tokenList;
+  tokenListCopy.push_back(nextToken);
+
+  // Create a empty node of tokens from radix tree.
+  std::shared_ptr<NodeData> evictedNodeData = nullptr;
+  std::shared_ptr<NodeData> nodeData =
+      this->rootTree->Insert(tokenListCopy, evictedNodeData);
+  if (nodeData == nullptr) {
+    LOG(ERROR) << "insert failed";
+    return;
+  }
+  KVStateCacheBlockBuilder* kvStateCacheBlockBuilder =
+      reinterpret_cast<KVStateCacheBlockBuilder*>(
+          (reinterpret_cast<TreeData*>(nodeData->treeData->data))
+              ->kvStateCacheBlockBuilder);
+  if (evictedNodeData != nullptr) {
+    Delete(evictedNodeData);
+  }
+
+  if (kvStateCacheBlockBuilder->IsFull()) {
+    /**
+     * If the kv-state cache of the tree is full, trigger split. Delete the
+     * empty node from the radix tree and split the tree. Then, kv-state cache
+     * split according to the new tree.
+     */
+    VLOG(100) << "trigger splits";
+    std::shared_ptr<NodeData> evictedNodeData = nullptr;
+    this->rootTree->Delete(tokenListCopy, evictedNodeData);
+
+    std::shared_ptr<NodeData> subTreeHeader;
+    std::vector<std::shared_ptr<NodeData>> nodeDataList =
+        rootTree->Split(tokenListCopy, subTreeHeader);
+    KVStateCacheBlockBuilder* newKVStateCacheBlockBuilder =
+        Split(client, kvStateCacheBlockBuilder, nodeDataList);
+
+    TreeData* newTreeData = new TreeData();
+    newTreeData->kvStateCacheBlockBuilder = newKVStateCacheBlockBuilder;
+    newTreeData->isPtr = true;
+
+    subTreeHeader->treeData->data = newTreeData;
+    subTreeHeader->treeData->dataLength = sizeof(TreeData);
+    rootTree->SetSubtreeData(newTreeData);
+    VLOG(100) << "block split success";
+
+    // kv_state_cache_builder->UnLock();
+    Update(client, tokenList, nextToken, kvState);
+  } else {
+    // Update the kv-state cache.
+    OffsetData* data = new OffsetData();
+    kvStateCacheBlockBuilder->Update(kvState, data);
+    nodeData->nodeData->data = data;
+    nodeData->nodeData->dataLength = sizeof(OffsetData);
+  }
+
+  VLOG(100) << "builder:" << kvStateCacheBlockBuilder
+            << " bitmap:" << kvStateCacheBlockBuilder->GetBitmapStr();
+}
+
+static std::shared_ptr<NodeData> node;
```

Comment:
Why a global static variable is needed?

Reply:
This should have been removed earlier. 
I remove it now.